Adding parameters as a concept to stream/hal/tooling.

benvanik opened the pull request:
Parameters are externalized storage for resources that are asynchronously accessible and device-aware. Parameters can be read or written on the same device timelines as the operations that consume or produce them and with locality pinning to ensure memory doesn't need to move. Parameters are referenced by an optional scope (a file name, a model name, whatever) and a unique key within that scope, with the scope being strongly recommended as a way to distinguish sets of parameters that may exist when multiple model parts are compiled together and would otherwise collide.

Parameters are provided to programs by a virtual interface and can support shared parameters (same storage used in multiple contexts, or outliving a single instantiation in a context), in-memory caches, memory-mapped files (including directly using the mapped memory for execution when devices support it), iree_hal_file_t usage for device-supported I/O, and parameter subsetting for things like runtime sharding. A basic file cache is implemented to allow for programs to decide when and where they want to use the parameters without needing to have bound them to devices at startup time.

Alongside read(+load) and write operations gather and scatter allow for batching of large numbers of reads and writes into/from single buffers. For parameter providers that can batch operations this allows for a handful (~1-4) of calls out to perform many more operations (~thousands). Modeling the gather/scatter also gives us a point where we could extract the mapping and use it to repack files/defrag memory in the future.

Parameters are currently defined by the `#stream.parameter.named` attribute which specifies an optional parameter scope and a scope-unique key for the parameter along with its logical type. Today these are intended to be used as default values on global ops (mutable or immutable) and are hackily processed as if they were constants. Future changes will allow parameter mutation and storage but what's present should be enough for inference and training parameter initialization.

Example parameter (here a tensor but parameters can be other types in the future to act as bags of bits):
```mlir
util.global private @"model.layer-1.kernel" = #stream.parameter.named<"mnist"::"model.layer-1.kernel"> : tensor<784x128xf32>
```

Parameters can optionally have a subrange specified indicating that the logical tensor is a block of some larger storage. When sharding this can be used to have an individual shard load a subset of the parameter data:
```mlir
util.global private @"model.layer-1.kernel-shard-0" = #stream.parameter.named<"mnist"::"model.layer-1.kernel", {offset = 0}> : tensor<392x128xf32>
util.global private @"model.layer-1.kernel-shard-1" = #stream.parameter.named<"mnist"::"model.layer-1.kernel", {offset = 200704}> : tensor<392x128xf32>
```

In this initial implementation we err on the side of optimizing for discrete memory devices (GPUs/etc) by emitting gathers of all parameters. On unified memory systems where we can zero-copy import parameters into device memory this is wasteful but it ensures proper alignment/packing/limited runtime overheads. Setting the resource memory model to `unified` via the `#stream.resource_config` attribute (helper flag `--iree-stream-resource-memory-model=unified`) will change instead to aliasing parameter memory where possible at the cost of increased runtime overhead. Future changes will connect the resource memory model to those of the devices under compilation and allow for heterogenous deployments to treat parameters used exclusively on different devices in whatever way is best for that device.

Basic tooling support for read-only parameters has been added for testing by allowing parameter files to be specified on the command line:
```
$ iree-run-module \
    --parameter_mode=mmap \
    --parameters=some_scope=some/file0.safetensors \
    --parameters=other_scope=some/file1.gguf \
    --module=...
```

Currently parameters are only usable from the full HAL implementation and not the inline HAL - the parameter file format and index code was kept portable such that it could be reused for a lighter-weight feature set if we wanted to support parameters in the inline HAL but given that cases where the inline HAL is interesting are usually small models on tiny systems where optimization of parameters is critical to memory/performance I haven't bothered here.

Since all parameter file formats are terrible a new parameter file format that is less terrible for our uses will be introduced in future changes. It's still experimental and not fully wired up but will be something we can convert other formats into for optimize use as both immutable constant and mutable variable storage in our tools when direct compatibility with existing frameworks is not required without conversion steps.

The `iree-dump-parameters` tool can be used to inspect any of the parameter file formats the tooling can load and extract individual parameters. It indexes parameters using the same flags as the rest of the tooling so it can also be useful to see what parameters are actually available for use without trial and error in other tools. Example output:
```
$ ../iree-build/tools/iree-dump-parameters.exe --parameters=a=tools/test/parameters_a.safetensors --parameters=runtime/src/iree/io/formats/gguf/testdata/multiple.gguf --extract=a::a0=a0.bin --extract=tensor0=tensor0.bin
//===--------------------------------------------------------------------------------------------------------------===//
// Parameter scope `a` (2 entries, 64 total bytes)
//===------------+------------------+------------------+-----------------------------------------------------------===//
//         Start |              End |           Length | Key
//---------------+------------------+------------------+--------------------------------------------------------------//
             120 |              152 |               32 | `a0`
             152 |              184 |               32 | `a1`

//===--------------------------------------------------------------------------------------------------------------===//
// Parameter scope `` (3 entries, 72 total bytes)
//===------------+------------------+------------------+-----------------------------------------------------------===//
//         Start |              End |           Length | Key
//---------------+------------------+------------------+--------------------------------------------------------------//
             448 |              464 |               16 | `tensor0`
             512 |              520 |                8 | `tensor1`
             576 |              624 |               48 | `tensor2`

Extracting parameter `a::a0` (32b) to `a0.bin`...
Extracting parameter `tensor0` (16b) to `tensor0.bin`...
```

Progress on #14987.

benvanik committed:
Adding parameters as a concept to stream/hal/tooling.
Parameters are externalized storage for resources that are
asynchronously accessible and device-aware. Parameters can be read or
written on the same device timelines as the operations that consume or
produce them and with locality pinning to ensure memory doesn't need to
move. Parameters are referenced by a scope and a key, with the scope
being optional but strongly recommended as a way to distinguish sets of
parameters that may exist when multiple model parts are compiled together
and would otherwise collide.

Parameters are provided by a few operations implementing a virtual
interface and can support shared parameters (same storage used
in multiple contexts, or outliving a single instantiation in a context),
in-memory caches, memory-mapped files (including directly using the
mapped memory for execution when devices support it), iree_hal_file_t
usage for device-supported I/O, and parameter subsetting for things
like runtime sharding.

Alongside read(+load) and write operations gather and scatter allow for
batching of large numbers of reads and writes into/from single buffers.
For parameter providers that can batch operations this allows for a
handful (~1-4) of calls out to perform many more operations (~thousands).
Modeling the gather/scatter also gives us a point where we could extract
the mapping and use it to repack files/defrag memory in the future.

Basic tooling support for read-only parameters has been added for testing
by allowing parameter files to be specified on the command line:
```
$ iree-run-module \
    --parameters=some_scope=some/file0.safetensors \
    --parameters=other_scope=some/file1.gguf \
    --module=...
```

Since all parameter file formats are terrible a new parameter file
format that is less terrible for our uses will be introduced in future
changes. It's still experimental and not fully wired up but will
be something we can convert other formats into for optimize use as both
immutable constant and mutable variable storage in our tools when direct
compatibility with existing frameworks is not required without conversion
steps.

Progress on #14987. Need an attribute and stream constant/variable
packing to emit the new parameter ops.

stellaraccident commented on runtime/src/iree/io/formats/gguf/gguf_format.c:
```diff
@@ -0,0 +1,641 @@
+
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
```
Ugh. This is... not a good format.

benvanik replied:
Yeah - does a few things right but needs a few more iterations. I'm not sure they have the right pressure to make it better, though, and I suspect it'll just get more weird stuff shoved into it. That you need the full format to parse the file but their own code #ifdefs some of the info based on compiler flags shows me they haven't really worked through compatibility. At least it has alignment :P

stellaraccident commented on runtime/src/iree/io/formats/gguf/testdata/generate_gguf_files.py:
```diff
@@ -0,0 +1,53 @@
+# Lint as: python3
```
We really don't need this anymore.

stellaraccident commented on runtime/src/iree/io/formats/safetensors/safetensors_format.c:
```diff
@@ -0,0 +1,520 @@
+
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
```
If I had a dollar for every time I'd written something like this... I'd have about $3.23. Good job.

stellaraccident reviewed (approved):
I feel like whenever I review one of these monsters, I need to declare that I did indeed read and review every line and double checked allocation/indexing.

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/Conversion/StreamToParams/Patterns.h:
```diff
@@ -0,0 +1,23 @@
+namespace mlir::iree_compiler {
+
+// Populates conversion patterns for stream->HAL (loader).
+void populateStreamToIOParametersPatterns(MLIRContext *context,
```
Update comment (or just remove, function name is self-explanatory)

```suggestion
void populateStreamToIOParametersPatterns(MLIRContext *context,
```

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/IR/IOParametersDialect.cpp:
```diff
@@ -0,0 +1,99 @@
+struct IOParametersInlinerInterface : public DialectInlinerInterface {
+  using DialectInlinerInterface::DialectInlinerInterface;
+
+  bool isLegalToInline(Operation *call, Operation *callable,
+                       bool wouldBeCloned) const final {
+    // Sure!
+    return true;
```
Sure! :D

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/IR/CMakeLists.txt:
```diff
@@ -0,0 +1,81 @@
+    --gen-op-defs IOParametersOps.cpp.inc
+)
+
+iree_tablegen_doc(
+  NAME
+    IOParametersDialectDocGen
+  TD_FILE
+    "IOParametersOps.td"
+  OUTS
+    --dialect=io_parameters --gen-dialect-doc IOParametersDialect.md
```
(If you'd like: I can do this as a follow-up, or I can send a PR to this branch)

For this to show up on the website (https://iree.dev/reference/mlir-dialects/), it should be added to the index page (https://github.com/openxla/iree/blob/main/docs/website/docs/reference/mlir-dialects/index.md) and nav: https://github.com/openxla/iree/blob/573f5e91901ce54c594b0e82f894a98f9f9d53f7/docs/website/mkdocs.yml#L150-L164

See also:
* https://github.com/openxla/iree/blob/main/docs/website/generate_extra_files.sh
* https://github.com/openxla/iree/blob/main/docs/website/postprocess_dialect_docs.py

ScottTodd replied:
https://github.com/openxla/iree/pull/15405 if you want it

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/IR/IOParametersBase.td:
```diff
@@ -0,0 +1,71 @@
+    asynchronously accessible and device-aware. Parameters can be read or
+    written on the same device timelines as the operations that consume or
+    produce them and with locality pinning to ensure memory doesn't need to
+    move. Parameters are referenced by a scope and a key, with the scope
+    being optional but strongly recommended as a way to distinguish sets of
+    parameters that may exist when multiple model parts are compiled together
+    and would otherwise collide.
```
I'd like an example or diagram useful to understand "when multiple model parts are compiled together and would otherwise collide" (a general example of "multiple model parts compiled together" would be nice too, independent of the comments here)

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/IR/IOParametersOps.td:
```diff
@@ -0,0 +1,237 @@
+]> {
+  let summary = [{reads a parameter from a parameter scope}];
+  let description = [{
+    Asynchronously reads a parameter from an external parameter provider and
+    returns the resulting buffer. Depending on the parameter and buffer types
```
Just a thought while still early in the review: can we see parameter loads in Tracy? I'm thinking of the network panel in Chrome Dev Tools as the sort of UI I might want (though gather/scatter could complicate that?)

benvanik replied:
Yep! Everything is instrumented so we can see individual parameters getting read/written with their keys associated with the zones.

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/IR/IOParametersOps.td:
```diff
@@ -0,0 +1,237 @@
+  let arguments = (ins
+    HAL_Device:$device,
+    HAL_DeviceQueueAffinity:$queue_affinity,
+    HAL_Fence:$wait_fence,
+    HAL_Fence:$signal_fence,
```
So nice to see wait/signal fences on all these ops, since that gives us pipelining and synchronization naturally :)

benvanik replied:
yeah! makes it not usable from the inline hal but that's ok - an inline parameters module would be simpler/smaller and could still reuse all the file format parsing stuff (just not the parameter_provider interface that is async-only)

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/IR/IOParametersOps.cpp:
```diff
@@ -0,0 +1,391 @@
+// custom<ParameterReference>($scope, $key)
+//===----------------------------------------------------------------------===//
+
+static ParseResult parseParameterReference(OpAsmParser &parser,
```
Do you want a test file for these printers/parsers? I don't see one.

benvanik replied:
they're covered by the conversion tests (which have all the param ops) - for consistency I'll create the standalone op tests though

ScottTodd commented on compiler/src/iree/compiler/Modules/IO/Parameters/Conversion/ParamsToVM/test/parameter_ops.mlir:
```diff
@@ -0,0 +1,134 @@
+  // CHECK: %[[TARGET_BUFFER:.+]] = vm.call @io_parameters.load
+  // CHECK-SAME: (%[[DEVICE]], %[[QUEUE_AFFINITY]], %[[WAIT]], %[[SIGNAL]],
+  // CHECK-SAME:  %[[SCOPE]], %[[KEY]], %c50, %c48, %c527363, %c100)
+  %target_buffer = io_parameters.load<%device : !hal.device> affinity(%queue_affinity) wait(%wait) signal(%signal) source("scope"::"key")[%c50_i64] type("DeviceVisible|DeviceLocal") usage("TransferSource|TransferTarget|Transfer|DispatchStorageRead|DispatchStorageWrite|DispatchStorage|SharingImmutable") : !hal.buffer{%c100}
```
Would line breaks make sense in this op printer? This line is very long

Maybe like this? (Can't do much about the usage enums besides add collections of enums...?)

```mlir
%target_buffer = io_parameters.load<%device : !hal.device> {
  affinity(%queue_affinity)
  wait(%wait)
  signal(%signal)
  source("scope"::"key")[%c50_i64]
  type("DeviceVisible|DeviceLocal")
  usage("TransferSource|TransferTarget|Transfer|DispatchStorageRead|DispatchStorageWrite|DispatchStorage|SharingImmutable")
} : !hal.buffer{%c100}
```

(similar question for the other `io_parameters` ops)

benvanik replied:
at the point in the IR where these appear there's lots of long stuff, so it's fine - most investigation will happen at the stream level where things are nicely formatted/terse (these are just alongside all the other hal ops that look the same)

ScottTodd reviewed (commented):
Reviewed the `compiler/` changes. Looking at `runtime/` now.

ScottTodd commented on runtime/src/iree/io/formats/gguf/gguf_format.h:
```diff
@@ -0,0 +1,29 @@
+// Parses a .gguf file and merges its contained resources into |index|.
+//
+// Specification:
+// https://github.com/philpax/ggml/blob/gguf-spec/docs/gguf.md
```
This link 404s. https://github.com/ggerganov/ggml/blob/master/docs/gguf.md is more recent (hopefully no breaking changes)

ScottTodd commented on runtime/src/iree/io/formats/gguf/gguf_format_test.cc:
```diff
@@ -0,0 +1,105 @@
+  iree_io_parameter_index_release(index);
+}
+
+TEST(GgufFormatTest, SingleTensors) {
```
nit: singular/plural?
```suggestion
TEST(GgufFormatTest, SingleTensor) {
```

ScottTodd commented on runtime/src/iree/io/formats/gguf/testdata/generate_gguf_files.py:
Thanks for the generator script. Nice and simple too :D

ScottTodd commented on runtime/src/iree/io/formats/gguf/gguf_format.c:
```diff
@@ -0,0 +1,641 @@
+#include <ctype.h>
+
+// File format:
+// https://github.com/philpax/ggml/blob/gguf-spec/docs/gguf.md
```
Outdated, replace with https://github.com/ggerganov/ggml/blob/master/docs/gguf.md ?

ScottTodd commented on runtime/src/iree/io/formats/gguf/gguf_format.c:
```diff
@@ -0,0 +1,641 @@
+//
+// References:
+// https://github.com/ggerganov/ggml/blob/master/src/ggml.c#L21213
+// https://github.com/ggerganov/llama.cpp/blob/master/gguf-py/gguf/gguf.py#L639
```
These URLs linking to specific lines are not stable. Convert to permalinks?
![image](https://github.com/openxla/iree/assets/4010439/332ae6e3-7086-448b-bec7-94b02db51057)
e.g. https://github.com/ggerganov/ggml/blob/ff904332ec131bfa3bd45ac552d2f84010d97e02/src/ggml.c#L19971-L19978

ScottTodd commented on runtime/src/iree/tooling/CMakeLists.txt:
```diff
@@ -165,6 +166,29 @@ iree_cc_test(
+    iree::base::internal::flags
+    iree::base::internal::path
+    iree::hal
+    iree::io::formats::gguf
+    iree::io::formats::safetensors
```
These formats aren't always included in the runtime, right? Looks like just added as deps to tools and then users can include if they want them?

benvanik replied:
correct - base runtime doesn't pull in the formats - not even the iree/io/ tree or the parameters module - so bare metal and such don't need any of this

benvanik committed:
Adding iree-dump-parameters tool.
Supports enumerating the parameters in any parameter file we support
and extracting parameter contents to individual binary files.

Example output:
```
$ ../iree-build/tools/iree-dump-parameters.exe --parameters=a=tools/test/parameters_a.safetensors --parameters=runtime/src/iree/io/formats/gguf/testdata/multiple.gguf --extract=a::a0=a0.bin --extract=tensor0=tensor0.bin
//===--------------------------------------------------------------------------------------------------------------===//
// Parameter scope `a` (2 entries, 64 total bytes)
//===------------+------------------+------------------+-----------------------------------------------------------===//
//         Start |              End |           Length | Key
//---------------+------------------+------------------+--------------------------------------------------------------//
             120 |              152 |               32 | `a0`
             152 |              184 |               32 | `a1`

//===--------------------------------------------------------------------------------------------------------------===//
// Parameter scope `` (3 entries, 72 total bytes)
//===------------+------------------+------------------+-----------------------------------------------------------===//
//         Start |              End |           Length | Key
//---------------+------------------+------------------+--------------------------------------------------------------//
             448 |              464 |               16 | `tensor0`
             512 |              520 |                8 | `tensor1`
             576 |              624 |               48 | `tensor2`

Extracting parameter `a::a0` (32b) to `a0.bin`...
Extracting parameter `tensor0` (16b) to `tensor0.bin`...
```

ScottTodd commented on runtime/src/iree/io/formats/gguf/gguf_format_test.cc:
/me adds to list of tests to add fuzzers for at some point

benvanik replied:
one day we'll have fuzzers, I hope :P

ScottTodd commented on runtime/src/iree/io/formats/safetensors/safetensors_format.h:
```diff
@@ -0,0 +1,41 @@
+// to use this is if sourcing from a Hugging Face model that has its weights
+// already in the safetensors format.
+//
+// WARNING: this implementation has not been thoroughly tested or verified as
+// safe or correct. Use with caution only on trusted inputs. Tip: don't embed
```
/me adds to list of tests to add fuzzers for at some point

(edit: read the _format.c... oh... hmm 🤔)

ScottTodd commented on runtime/src/iree/io/formats/safetensors/testdata/generate_safetensors_files.py:
```diff
@@ -0,0 +1,31 @@
+# https://huggingface.co/docs/safetensors/index
+#
+# To regenerate:
+#  $ pip install torch safetensors
```
could also put this in a requirements.txt file, possibly pinned to a known-working version if unstable

(fine to ignore, I don't think we'll have enough people updating these regularly enough to need more than this comment)

benvanik replied:
yeah I don't want to pull in big/shady deps - my hope is these never change unless the file format changes so only one person will ever need to install them/rerun this

ScottTodd replied:
Ah, I mean a requirements.txt _in this directory_, like https://github.com/openxla/iree/blob/main/docs/website/requirements.txt. That lets people install local requirements from any directory without needing to read through the files to see what is needed, and also lets us pin versions or specify minimums. Individual files would not affect the regular build of the python runtime or general requirements for using IREE.

ScottTodd commented on runtime/src/iree/io/parameter_provider.h:
```diff
@@ -0,0 +1,241 @@
+  // Providers should drop any transient memory that is possible to reallocate
+  // upon resume.
+  IREE_IO_PARAMETER_PROVIDER_SIGNAL_SUSPEND = 1,
+  // Program has received a low memory alert.
+  // Providers must aggressively drop all possible memory even if expensive to
+  // rematerialize it. On some platforms this is sent as a threat that if
+  // sufficient memory is not unwired/freed ASAP the process will be killed.
+  IREE_IO_PARAMETER_PROVIDER_SIGNAL_LOW_MEMORY = 2,
```
And on some platforms, the process will be killed anyway 🙃

benvanik replied:
/me eyes ios

ScottTodd commented on runtime/src/iree/tooling/parameter_util.c:
```diff
@@ -0,0 +1,276 @@
+    "\n"
+    "Supported formats:\n"
+    "- .irpa (IREE parameter archive)\n"
+    "- .gguf (https://github.com/philpax/ggml/blob/gguf-spec/docs/gguf.md)"
```
update link: https://github.com/ggerganov/ggml/blob/master/docs/gguf.md
```suggestion
    "- .gguf (https://github.com/ggerganov/ggml/blob/master/docs/gguf.md)"
```

ScottTodd commented on runtime/src/iree/tooling/parameter_util.c:
```diff
@@ -0,0 +1,276 @@
+    "`my_scope=some_file.gguf`.\n"
+    "\n"
+    "Supported formats:\n"
+    "- .irpa (IREE parameter archive)\n"
```
nit: Could leave `.irpa` off until it is implemented

benvanik replied:
/me types furiously

stellaraccident committed:
Appease enforce glob.
